### ISSUE TITLE
timeoutson initial sync

### ISSUE THREAD

**Issue by delendor** [jobisoft/TbSync#34]
I recently searched for a way to connect a new android phone to my Zarafa (predecessor of Kopano) server and found tbsync which was a bit tricky to configure at first (I tried via "New calendar" from Lightning Plugin, but didn't have an account setup), but then found the 'account manager'. I added an account and immediately synced everything nicely. 

When I tried adding a second account things went from "not quite" to "never again". 

The second account (on the same server) has a bigger address book, a much longer calendar history and a way bigger email database (but that isn't synced via tbsync). 

Although I haven't configured anything special into the bigger account, the sync just wouldn't finish. it always runs into one issue or the other. I exclusively use the https feature and the server possesses a valid (letsencrypt) certificate, using a subdomain (like in delendor.domain.com). 

When I tried to set up tbsync on a computer at work (also Linux OS), I found myself able to correctly sync the address book of the small account before it then timeouts at some stage. Retries didn't help. The bigger account -as experienced at home- never synced anything.

I even tried checking the logs (both server and client but to not much avail. The initial handshake seems to get across, but then things slow down until the client gives up. The server might be a bit slowish. It's a synology NAS (almost 7 years old) with more and more apps running in parallel. 

What do I do to properly diagnose th problem?

**Comment by jobisoft** [comment]
The first account, which worked alone, also worked with the second account setup? Just the second account fails?

What happens, if you disconnect the first account and sync only the second (larger) one? Maybe disconnect (and reconnect) that one also to have a clean restart.

Please enable the debug log (under Help in the account manager), if that log does not help you to debug the problem, it would like to take a look.

**Comment by delendor** [comment]
As far as I remember, yes. The first one was ok. I could add and remove appointments and they would quickly appear on the mobile phone (which was set-up to sync as well via exchange protocol). I also have a second "small" user, which I could connect to and sync ok. Only the big one never worked out. I tried diconnecting and reconnecting quite often, with no luck. I also started completely from scratch, deleted the account and created it anew. Same problem. 
When I checked just now, I ...
a) found out about the second small WORKING one (which I had actually forgotten before)
b) that the first one (that completely synced initially) has issues now. As I'm writing this, it still tries to sync, but strangely doesn't timeout as quickly as the big one. 
c) found that this thunderbird now has three accounts, one that made its initial sync, but now shows a yellow triangle with exclamation, the second is still green, apparently ok, although I never added data through thunderbird to that. The third is still red circle, white exclamation mark, as it never even got one complete sync. All accounts are on the same server, which I don't connect to using its local IP (as this might trigger certificate errors) but through its domain name, which I link to the server by /etc/hosts entry such that there is a direct connection being made.

**Comment by jobisoft** [comment]
if you click throu the accounts, you get a SyncStatus for each. Can you give me the status of each of your accounts? Also the debug log will be very helpful.

**Comment by delendor** [comment]
I activated debug log, restarted thunderbird to clear the log, then start TB, go to account manager, make screenshot. After I made that screen shot, I pushed syncronize all, which timed out on the third (big) account, but worked ok for the first two again. (I did give up after minutes previously from the same room, but laptop in a different position, so wireless reception may have been slightly different). Sending you the debug log from this "sync all" action.
After this, the sync status for the first two was ok, the last (ZZZZ) has "Communication timeout"

[debug_sanitized.log](https://github.com/jobisoft/TbSync/files/1648036/debug_sanitized.log)
![tbsync-screenshot-2018-01-19](https://user-images.githubusercontent.com/25579643/35173693-93e017da-fd6c-11e7-948e-2b2b351be1ec.jpg)

**Comment by delendor** [comment]
Now after reconnecting the first account and trying to sync (same position of laptop) it takes forever (at least 10 minutes, still at synchronizing contacts, which didn't even change anywhere).

**Comment by jobisoft** [comment]
The log does not contain a timeout error and it only contains the sync log of two accounts, probably the two accounts that did not fail.

I saw that tzpush is installed. TbSync contains tzpush code, so I do not know, if crazy things happen, if both are installed/enabled. 

But: Timeout Errors are bad. If I get a javascript or protocol error , I know I did something wrong and can fix it. There is no feedback whatsoever with timeouts. The only thing I saw is this:

[JavaScript Error: "$$$$$$$$$$$$$$$$$$$ : server does not support RFC 5746, see CVE-2009-3555"]

I would suggest the following:
 - only test with one account for now, a small one
 - test with tbsync only (tzpush disabled)
 - test without https (its inside you own network, right?), you may have to "unlock" the default settings, if you used autodiscover

Just to get an idea, of what could go wrong . If all that did not help, does it work with tzpush?

**Comment by jobisoft** [comment]
My last post was not correct. There is a initial folderrequest for ZZZZ, which does not get an answer. 

Can you sync that account with another client via Exchange Active Sync? For example with your android? or is it just TbSync, who is failing?

**Comment by jobisoft** [comment]
Furthermore: The blue sync arrows rotate everytime something is send to or received from the server. If they stand still for longer than 30s, something is really wrong, because sync should stop with a timeout error and the sync arrows should be replaced by a red exclamation mark - it should not "hang".

**Comment by delendor** [comment]
Syncing the large account has been absolutely flawless from iPad / iPhone. I did get a few Sync error messages from the Android today, but after a while, the new entry appeared on the server, so apparently that device could recover. 
The latest sync attempt ended in "contacts synced ok, calendar communication timeout, tasks Not syncronized", so since this is a wifi-ed laptop, I guess depending on network quality I get either good sync or nothing. 
What crushes my spirits is the fact that on the desktop computer, I operate a GBit interface to the router which is directly connected to the NAS server, so there shouldn't be any issue with the connection, but still there is...

This was my impression. I gave up (close TB) when it took too long without anything happening. 

If there's anything to crank up debug level(s), let me know.

**Comment by jobisoft** [comment]
Could you re-check the blue arrow rotation, that it does not "hang" for more than (about) 30s and will eventually return to an error state? Also the Indicator in the lower right corner should switch back from "TbSync: .:...." to "TbSync: Error".  (Everytime the arrow rotates, the indicator dot moves as well) 

Something like "It is syncing for more than 10 Minutes without progress" should not happen.

The only possible way for a sync interruption to occur is TbSync asking for a password and the password prompt for some reasons pops up in the background, but this should be viewable with ALT+TAB.

Did turning of https make any difference?

Does it work with tzpush?

Does it work with TbSync [v0.6.8.2](https://github.com/jobisoft/TbSync/releases/download/v0.6.8.2/TbSync-0.6.8.2.xpi)

I will include an option in the next release to manually set the timeout, so you could increase it (maybe your server just needs longer...)

**Comment by delendor** [comment]
The rotating arrow was clearly longer than 10 minutes but at least every time I left it alone it returned with an error, but I didn't try to see when onky returned after some hours. Since the machine has 3 accounts maybe it did after 30 minutes?

Https vs http makes no difference.

Tzpush addon was an old installation and probably didn't work hence I looked for something different now and found tbsync

So indeed the server might play an important role in this problem. 

Tbsync v0.6.8.2 is up next

**Comment by jobisoft** [comment]
so I do not forget other things to ask you:
 - test on windows possible?
 - test with other EAS version possible? (what are you using now?)

**Comment by jobisoft** [comment]
I released a beta, which allows to set timeout:
https://github.com/jobisoft/TbSync/releases/download/v0.6.9.1/TbSync-0.6.9.2-beta1.xpi

Go to thunderbird options -> Advanced -> Edit Config and search for tbsync. You will get en entry for

extensions.tbsync.eas.timeout

which defaults to 30000 (ms) which are 30s. Try to set it higher and see if it makes any difference.

**Comment by delendor** [comment]
Much better now. The account which originally synced ok and later refused (time.out) occasionally, typically on calendar entries (after syncing contacts ok) went through nicely after setting timeout to 60sec.
The second (very small) account went fine as well. 
The large one is still syncing, still contacts, it's taking more than 15 minutes already (will keep it running until tomorrow if need be) just to see. 
I guess, I have a good argument for a new NAS now, I wanted a backup station anyway... ;-)

**Comment by delendor** [comment]
Test on windows would be possible, although I have no need for windows anymore (made the switch to Linux when XP ended its supported life). 
I'm guessing I could test many things, but what exactly is EAS. 

One thing I did (before installing the beta) was removing yet another EWS provider plugin, which also didn't do the job at the time, but that was never installed on the other machine, so it's probably unrelated.

**Comment by jobisoft** [comment]
Ok, Sounds good. 

For the big account currently syncing contacts for over 15min: You could check the address book, if something is syncing (number of contacts should increase over time)



Am 21.01.18 um 17:30 schrieb delendor

**Comment by jobisoft** [comment]
@windows: Since it is working now, this is low priority...

@EAS: This is short for Exchange Active Sync, the protocol TbSync is using to sync. EAS has different protocol versions, TbSync supports 2.5 and 14.0. If you disconnect your account, you can see the version you are currently using. It might be interesting to see what happens, if you use the other one.

**Comment by jobisoft** [comment]
I think, there is also no need to test v0.6.8.2 anymore...

**Comment by jobisoft** [comment]
If you are going to test against EAS v2.5, please use beta2, which I just [released](https://github.com/jobisoft/TbSync/releases/download/v0.6.9.1/TbSync-0.6.9.2-beta2.xpi)

**Comment by delendor** [comment]
With the latest progress, I wanted to try a remote sync from my work PC (also linux). I installed the beta-2 which I found at github, but the plugin wouldn't start. 
Tried to revert to the latest release. 
Same problem. 

this is from the debug.log:
** Mon Jan 22 2018 10:45:51 GMT+0100 (CET) **
[JavaScript Error: "A promise chain failed to handle a rejection. Did you forget to '.catch', or did you forget to 'return'?
See https://developer.mozilla.org/Mozilla/JavaScript_code_modules/Promise.jsm/Promise

Date: Mon Jan 22 2018 10:45:44 GMT+0100 (CET)
Full Message: Unknown account!
Thrown by db.getAccount(4)
Full Stack: JS frame :: resource://gre/modules/Promise.jsm -> resource://gre/modules/Promise-backend.js :: PendingErrors.register :: line 194
JS frame :: resource://gre/modules/Promise.jsm -> resource://gre/modules/Promise-backend.js :: this.PromiseWalker.completePromise :: line 715
JS frame :: resource://gre/modules/Task.jsm :: TaskImpl_handleException :: line 456
JS frame :: resource://gre/modules/Task.jsm :: TaskImpl_run :: line 331
JS frame :: resource://gre/modules/Promise.jsm -> resource://gre/modules/Promise-backend.js :: Handler.prototype.process :: line 935
JS frame :: resource://gre/modules/Promise.jsm -> resource://gre/modules/Promise-backend.js :: this.PromiseWalker.walkerLoop :: line 813
JS frame :: resource://gre/modules/Promise.jsm -> resource://gre/modules/Promise-backend.js :: this.PromiseWalker.scheduleWalkerLoop/< :: line 747
JS frame :: resource://gre/components/nsPrompter.js :: openModalWindow :: line 364
JS frame :: resource://gre/components/nsPrompter.js :: ModalPrompter.prototype.openPrompt :: line 579
JS frame :: resource://gre/components/nsPrompter.js :: ModalPrompter.prototype.nsIPrompt_promptUsernameAndPassword :: line 770
JS frame :: resource://gre/components/nsPrompter.js :: ModalPrompter.prototype.promptAuth :: line 869
JS frame :: resource://gre/components/nsPrompter.js :: Prompter.prototype.promptAuth :: line 108
JS frame :: resource://gre/components/nsLoginManagerPrompter.js :: LoginManagerPrompter.prototype.promptAuth :: line 621
JS frame :: resource://calendar/modules/calAuthUtils.jsm :: cal.auth.Prompt.prototype.promptAuth :: line 275
JS frame :: resource://calendar/modules/calAuthUtils.jsm :: cal.auth.Prompt.prototype.asyncPromptAuth/promptlistener.onPromptStart :: line 304
JS frame :: resource://gre/components/msgAsyncPrompter.js :: runnablePrompter.prototype.run< :: line 28
JS frame :: resource://gre/modules/Task.jsm :: TaskImpl_run :: line 319
JS frame :: resource://gre/modules/Promise.jsm -> resource://gre/modules/Promise-backend.js :: Handler.prototype.process :: line 932
JS frame :: resource://gre/modules/Promise.jsm -> resource://gre/modules/Promise-backend.js :: this.PromiseWalker.walkerLoop :: line 813
JS frame :: resource://gre/modules/Promise.jsm -> resource://gre/modules/Promise-backend.js :: this.PromiseWalker.scheduleWalkerLoop/< :: line 747"]

**Comment by jobisoft** [comment]
You just hit another bug I am trying to fix. 

On that system, did you have an older version of TbSync running before installing the latest beta?

**Comment by delendor** [comment]
yes. probably the same as at home with a few accounts setup (I still expected at least one to work, but none of them did)

**Comment by delendor** [comment]
I can probably reproduce certain states, if you give me instructions.

**Comment by jobisoft** [comment]
Just to make sure we are talking about the same thing: You get the "Oops! TbSync was not able to start!" Message? 

Can you send me the entire debug log?



Am 22.01.18 um 11:09 schrieb delendor

**Comment by delendor** [comment]
beta3 starts (no more Oops message). 
I've configured my "big" account and started syncing. Timeout is set to 2 minutes! After about 1.5 h, I have many addresses in my TB address book, which definitely came from the server. So, something is working alright. I'll let you know, when the process finishes (I hope before I leave office), but unless it craps out at calendar items (as it used to), the issue seems to be solved and you may want to migrate the time-out setting into the user options for cases like mine with a sluggish server on a medium bandwidth connection (50 down / 10 up)

**Comment by jobisoft** [comment]
I would call this "sort of solved", because the large timeout does not have anything to do with your network connection. Exchange active sync was designed for mobile connections (3G) and does not need much bandwidth at all. 

There is still some unknown issue, why your server takes so long to answer my requests....

**Comment by jobisoft** [comment]
From this issue I also get the impression, that the feedback from the account manager about the current sync state could be improved.

Any suggestion what would have helped you to see, that TbSync is not hanging, but actually syncing, but very very slowly? The rotation of the arrows on each server response is obviously not enough, specially if the server needs very long to answer.

**Comment by delendor** [comment]
The rotation didn't help, indeed. The moving dot in the lower right corner (status bar) where tbsync account manager seems to sit, also did the same over and over. 
What you might show instead is every completion of one single item (contact with address, numbers and stuff or a single calendar entry) by counting up a variable and showing its content instead of rotating arrows. Or -if possible- the number inside the rotating arrows. 
If the number goes up, tbsync clearly still processes stuff. If you can get a number from the server about the total entries, it could even show xy / total, a pseudo progress bar. 
If it can't sync a single item within the set timeout, it is bound to fail. If it does, number goes up, user stays calm... ;-)

The unknown issue, why the server seems so slow, is actually understandable. It's a NAS box, 7 years old, an atom single-core CPU with 1 GB of memory, so it probably has to swap in and out database entries from disk. Zarafa's web interface also feels very sluggish and the "big" account has entries dating back almost the full period. I just checked and the free memory amounts to less than 100 MB. 

I agree that the indicator of a running synchronizaion could be improved, but the issue is clearly understood.

**Comment by jobisoft** [comment]
ok, I keep this open and work on improving the visual feedback for you to test. You have the only slow server I can test against :-)

**Comment by delendor** [comment]
It's interesting for me as well to see where the time is actually lost. I actually doubt that the data transfer is the bottle neck. Considering the latest finding with the work PC that had a lot of address book entries synced already and still took more than an hour (and wasn't finished when I called it a day, let's see tomorrow). It almost seems as if there is a diff against every single entry that takes just as long as the initial complete data transfer. But those things would get clearer with the mentioned indicator. Happy to keep testing...

**Comment by delendor** [comment]
Over night, the process ran on and on and finally into an error message. I attach the screenshot and the last 100 lines or so of the debug log. 
[final_lines_of_debug.log](https://github.com/jobisoft/TbSync/files/1655015/final_lines_of_debug.log)

Maybe the "shame" mentioned in the log messages applies to my server. The time of the error message coincides with the dial-up internet connection behind which my server sits being reset (actually accurate to the second), so I guess at least the message tells the truth. 

My current guess is that: If the process of syncinc does not complete within a certain time period [ one second (or one minute)], either the server or tbsync assumes changes (e.g. time stamp) and keeps repeating forever, until some other error stops it. I'll restart the sync now and report again.

**Comment by jobisoft** [comment]
I looked at the debug.log. I think you hit another bug. Any chance I can have to full debug log (via email)?


```
[Receiving data <requestingchanges> for privat (Contacts) (XML)] : 
<?xml version="1.0"?>
<Sync xmlns='AirSync'>
<Collections>
<Collection>
<SyncKey>{2ff19e79-27b0-4f48-b705-2a58e907c705}6</SyncKey>
<CollectionId>3067c8828f384fb4bad7e11799250524df0900000000</CollectionId>
<Status>1</Status>
</Collection>
</Collections>
</Sync>
```

This tells me, the sync has finished. However in the next line TbSync tells me:

`[Receiving cards] : Server told us, he has more cards. Requesting next set (would be a shame, if server sends them one by one).`

which means there was a MORE indicator, and I should request the next set of contacts, but there is no MORE indicator in the last response. So this looks like a self inflicted endless loop. (I added the "shame" comment because I was upset with a Kopano server which sends 250 contacts in 250 request-response-cycles, instead of 3 cycles in packs of 100/100/50 as I tell him to (WindowsSize in my requests). 

The actual error is 

`[finishFolderSync(privat.Contacts)] : Could not connect to server (NS_ERROR_NET_TIMEOUT).`

which means he lost connection to the server (because the connection was reset, as you said).

**Comment by delendor** [comment]
Today's resync went better. A comment about "local changes" in calendar was fixed with a (very quick) resync, now also the remote Thunderbird (at work) was able to sync using a timeout of 120000 ms and several hours for the initial sync and two resyncs. 

The initial sync apparently never finishes, unless stopped hard, although it basically did the job (since the timeout was relaxed). Maybe there is a bug in the "finish sync" step which somehow invalidates all received entries (or by sending them back to the server? which would mark them new there and repeat...? I'm just guessing here) and thereby restarting the process indefinitely. Or it tries to use a prior socket / handle, which has timed out after hours of entry syncing... ? 

If you like, I can try to double check, whether all accounts have this issue that the initial sync doesn't finish, but a later sync does...

**Comment by delendor** [comment]
The full debug log consisted of 3 or 30 million (no kidding) lines. I cleared it after rescuing the final battle. Sorry.
I did check just now, what actually was synced. And it seems that the full calendar database went across, including data reaching back to June 2013, so there WAS an enormous amount of data to be transferred.

**Comment by jobisoft** [comment]
Thanks for your feedback. There was indeed an endless loop in the contactsync code, if a certain condition was me. I just fixed that.

Could you please retry with [beta4](https://github.com/jobisoft/TbSync/releases/download/v0.6.9.1/TbSync-0.6.9.2-beta4.xpi)? Disconnect the account and reconnect, so we start from scratch....

**Comment by delendor** [comment]
Will do, but probably only tomorrow. Too much stuff in my inbox with "urgent" in the subject ATM.

**Comment by delendor** [comment]
beta4 synced from my home machine (GBit to the server) also fine. Didn't take very long. But: 30 seconds did NOT work, had several TIMEOUTs, then a certificate error due to local IP used but certificate only contains global domain name. After /etc/hosts fix and retry at timeout 120 sec, all worked fine. 
Tomorrow sometime, I'll check the remote sync from scratch.

**Comment by jobisoft** [comment]
Thanks! I just released [beta5](https://github.com/jobisoft/TbSync/releases/download/v0.6.9.1/TbSync-0.6.9.2-beta5.xpi), which includes some changes to the user feedback. This is the current state:

1. If the account manager is closed, the only feedback is the status indicator in the lower right corner, either "idle", "error" (which is idle but there was an error), and the "...:...." moving dot, indicating that "something" is going on (dot is not moving while waiting for server answers).

2. If the account manager is open, the user can see a detailed status of the currently selected account. I added much more "states"

old : receiving remote changes
new: requesting remote changes, waiting for remote changes, receiving remote changes

You will mostly see the "waiting" state, because thats what takes up most of the time: waiting for the server to answer. Furthermore, if it takes longer that 2s, the wait time will be added to the status and updated each second. Also, if a state is repeated (for example "waiting for remote changes", when not all changes where send in a single response), the number of repetitions is also added to the status. You could get something like this

```
waiting for remote changes (#4,5s)
```

Which means, this is the 4th time we have to request and wait for remote changes and we are already waiting 5s for the answer.

Is that working for you? I will also try to get info on how many items still have to be synced, instead of the plain repetition up-counter.

**Comment by jobisoft** [comment]
I did not like beta5, I reworked the user feedback for [beta6](https://github.com/jobisoft/TbSync/releases/download/v0.6.9.1/TbSync-0.6.9.2-beta6.xpi). Would you give it a try?

![image](https://user-images.githubusercontent.com/5830621/35328636-50c19fa2-00fd-11e8-8d37-1d9a44b00711.png)

The account status is counting down to the timeout now, starting if waiting for more then 2s.
The folder status is counting the actual elements being synced (if I was not able to get the total amount, it is just an element upcounter)

**Comment by delendor** [comment]
I like beta6, may I keep it ;-) 

Short report: 
I uninstalled the beta3, I was running before, which got the "first complete sync"  on the large account, albeit with an infinite loop to be stopped manually. Then installed beta6 and learned the new "progress bar" feature by just manually syncing, which it did. 
I then disconnected and reconnected, forcing a new sync from scratch. 
The folder list update took about 63 seconds of waiting (maybe there's a minute delay from the server side for some strange reason), after which the contacts started syncing. I could nicely follow the first 100/440 contacts which took about 20 seconds of waiting, as did the next 100 and the next 100 and the next 100. 8 seconds for the remaining 40 and 440 / 440 were synced ok. 
The calendar sync produced 338 total changes to be synced and it took about 22 seconds until the server responded, but the client actually froze at this point (system indicated unresponsive application, visible by darkening of window by window manager) for a few seconds, but came back, but with an error condition: NETWORK ERROR TIMEOUT or something like that, stopping the sync at this point. I restarted (without disconnect) and the rest went through ok with about 22 seconds on average for each 100 calendar items. Tasks list is empty, so the sync there wasn't possible to be followed. The unresponsiveness of the client (thunderbird) probably came about as the calendar plugin tried to go off with alarms on many (past) items and collected items, while they were still being added to the calendar for display in a popup window. This lead to a very high CPU usage and seemed as a stalled application to the system. Maybe, one should [if possible] disable alarms globally during a sync, or at least during an initial sync with more than (configurable) X items to be synced.

In summary: 
The progress indicator makes it much easier now when you face a slow server to understand that the sync is still operational. There was no more infinite looping, but a straight start to finish, although it took a while (just under 5 minutes). The spurious network error is nothing to worry about, IMHO. I do have some issues with my dial-up network at present as my DynDNS service told me that he lost connection several times during the day, this is unrelated to the tbsync process, but the reason for this might also affect the tbsync adversely.

Thank you !
Which paypal account did you say should I send money to ? :-)

**Comment by jobisoft** [comment]
Glad it is (almost) working for you now.

@disable old alarms: that is a very good idea, I will implement that in beta7
@paypal: https://www.paypal.me/johnbieling :-)

Thanks for your great feedback. I report back when beta7 is ready

**Comment by jobisoft** [comment]
Please try again with [beta7](https://github.com/jobisoft/TbSync/releases/download/v0.6.9.1/TbSync-0.6.9.2-beta7.xpi).

**Comment by delendor** [comment]
worked as expected. At home (with GBit instead of DSL 10 MBit connection), the initial folder list update was only a bit shorter as the remote one (still clearly violating the 30 seconds timeout, but not the 60 sec) but item sync (in blocks of 100) were about the same "waiting" time, ~20 sec per 100 items. The alarms don't pop up, which is good. I might be able to run one more "remote" sync from scratch, just to see, but this is not today.

**Comment by jobisoft** [comment]
Do plan to run a remote sync from scratch again? Otherwise I would like to close this issue.

**Comment by jobisoft** [comment]
Please update your beta version of TbSync to [0.7.6.8](https://github.com/jobisoft/TbSync/releases/download/v0.7.6/TbSync-0.7.6.8.xpi) or any later version available at the [release page](https://github.com/jobisoft/TbSync/releases).

This version now includes an update notification, so I no longer have to contact all beta users, to update their TbSync versions:

https://raw.githubusercontent.com/jobisoft/TbSync/master/screenshots/update_notification.png

While on the beta release channel, I strongly suggest to update, if a newer version is available. Once a new stable version gets released, you automatically switch back to the stable release channel (you no longer receive update notifications for beta versions).
